scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py

```diff
@@ -0,0 +1,153 @@
+from time import sleep
+
+"""
+The Maker class represents the coffee maker itself. It keep tracks of the water, milk,
+and beans in the machine. Makes coffee and reports on the resources.
+"""
+
+
+class Maker:
+
+    def __init__(self):
+        self.resources = {
+            "water": 300,
+            "milk": 200,
+            "beans": 100,
+        }
+
+    def report(self):
+        print(f"Water: {self.resources['water']}ml")
+        print(f"Milk: {self.resources['milk']}ml")
+        print(f"Coffee: {self.resources['beans']}g")
+
+    def is_resource_sufficient(self, ordered_menu_item):
+        can_make = True
+        for item in ordered_menu_item.ingredients:
+            if ordered_menu_item.ingredients[item] > self.resources[item]:
+                print(f"Sorry there is not enough {item}.")
+                can_make = False
+        return can_make
+
+    def make_coffee(self, ordered_menu_item):
+        for item in ordered_menu_item.ingredients:
+            self.resources[item] -= ordered_menu_item.ingredients[item]
+        print(f"Here's your {ordered_menu_item.name}. Enjoy! :)")
+
+
+"""
+The MenuItem class represents a menu item. It keeps track of the name, the ingredients,
+and the cost.
+"""
+
+
+class MenuItem:
+
+    def __init__(self, name, water, milk, beans, cost):
+        self.name = name
+        self.cost = cost
+        self.ingredients = {
+            'water': water,
+            'milk': milk,
+            'beans': beans,
+        }
+
+
+"""
+The Menu class represents the menu. It keeps track of the menu items where each item is
+an instance of the MenuItem class. The class holds the functionalities for finding an
+individual menu item and getting all the menu items.
+"""
+
+
+class Menu:
+
+    def __init__(self):
+        self.menu = [
+            MenuItem(name="latte", water=200, milk=150, beans=24, cost=2.5),
+            MenuItem(name="espresso", water=50, milk=0, beans=18, cost=1.5),
+            MenuItem(name="cappuccino", water=250, milk=50, beans=24, cost=3),
+        ]
+
+    def get_all_menu_items(self):
+        options = ""
+        for item in self.menu:
+            options += f"{item.name}/"
+        return options
```

Comment:
```suggestion
        return '/'.join(item.name for item in self.items)
```
Or, after adding `__str__` to `MenuItem`:
```suggestion
        return '/'.join(self.items)
```

- - -
Probably it also could be changed to `__str__` of the `Menu`.